atom/browser/api/atom_api_session.cc

```diff
@@ -330,6 +341,85 @@ void OnClearStorageDataDone(const base::Closure& callback) {
     callback.Run();
 }
 
+void DidReadBlobData(const scoped_refptr<net::IOBuffer>& blob_data,
+                     const Session::BlobDataCallback& completion_callback,
+                     int bytes_read) {
+  std::unique_ptr<base::BinaryValue> result(
+      base::BinaryValue::CreateWithCopiedBuffer(blob_data->data(),
+                                                bytes_read));
+  BrowserThread::PostTask(BrowserThread::UI, FROM_HERE,
+      base::Bind(&RunBlobDataCallback,
+                 completion_callback,
+                 base::Passed(&result)));
+}
+
+void DidCalculateBlobSize(
+    std::shared_ptr<storage::BlobReader> blob_reader,
+    const Session::BlobDataCallback& completion_callback,
+    int result) {
+  if (result != net::OK) {
+    std::unique_ptr<base::BinaryValue> dummy(new base::BinaryValue());
+    BrowserThread::PostTask(BrowserThread::UI, FROM_HERE,
+        base::Bind(&RunBlobDataCallback,
+                   completion_callback,
+                   base::Passed(&dummy)));
+    return;
+  }
+
+  uint64_t total_size = blob_reader->total_size();
+  int bytes_read = 0;
+  scoped_refptr<net::IOBuffer> blob_data =
+      new net::IOBuffer(static_cast<size_t>(total_size));
+  net::CompletionCallback callback = base::Bind(&DidReadBlobData,
+                                                base::RetainedRef(blob_data),
+                                                completion_callback);
+  storage::BlobReader::Status read_status = blob_reader->Read(
+      blob_data.get(),
+      total_size,
+      &bytes_read,
+      callback);
+  if (read_status != storage::BlobReader::Status::IO_PENDING)
+    callback.Run(bytes_read);
+}
+
+void GetBlobDataInIO(
+    const std::string& identifier,
+    Session::BlobIdType type,
+    content::ChromeBlobStorageContext* blob_context,
+    storage::FileSystemContext* file_system_context,
+    const Session::BlobDataCallback& completion_callback) {
+  std::unique_ptr<storage::BlobDataHandle> blob_data_handle;
+  if (type == Session::BlobIdType::PUBLIC_URL) {
+    blob_data_handle =
+        blob_context->context()->GetBlobDataFromPublicURL(GURL(identifier));
+  } else if (type == Session::BlobIdType::UUID) {
+    blob_data_handle =
+        blob_context->context()->GetBlobDataFromUUID(identifier);
+  }
+
+  if (!blob_data_handle) {
+    std::unique_ptr<base::BinaryValue> dummy(new base::BinaryValue());
+    BrowserThread::PostTask(BrowserThread::UI, FROM_HERE,
+        base::Bind(&RunBlobDataCallback,
+                   completion_callback,
+                   base::Passed(&dummy)));
+    return;
+  }
+
+  auto blob_reader = blob_data_handle->CreateReader(
+      file_system_context,
+      BrowserThread::GetMessageLoopProxyForThread(BrowserThread::FILE).get());
+  std::shared_ptr<storage::BlobReader>
```

Comment:
Moving the pointer from unique_ptr to shared_ptr usually means a bad design, you can add a helper class to put all blob reading methods there.